### ISSUE TITLE
Connect user's shell to a one-click-loaner

### ISSUE THREAD

**Issue by djmitche** [taskcluster/taskcluster-cli#54]
We have a fancy one-click-loaner option that is accessible via https://tools.taskcluster.net, but the terminal emulation is TERRIBLE and you can't copy/paste or do any of the normal terminal things.

Much better would be to connect my existing terminal to a one-click loaner:

```
dustin@lamport ~ $ taskcluster terminal $TASKID
# <--- shell prompt from loaner
```

This should support both websocket protocols
* https://github.com/taskcluster/docker-exec-websocket-server
* https://github.com/taskcluster/taskcluster-worker/tree/master/plugins/interactive/shellclient

**Comment by jonasfj** [comment]
taskcluster-worker already has a shell command implemented here:
https://github.com/taskcluster/taskcluster-worker/tree/master/commands/shell

That sets up the shell to work in raw mode, and tears it down when disconnecting.
Granted it might still have some edge cases around cleanup if disconnected incorrectly or so, I don't remember.

Obviously the shellclient can be reused from taskcluster-worker, for the docker-exec-websocket-server needs a client library reimplemented in golang.

**Comment by djmitche** [comment]
We have at least one user interested in this!

https://bugzilla.mozilla.org/show_bug.cgi?id=1335874

**Comment by jonasfj** [comment]
How interactive shell works right now:
 1. User creates a task with `task.payload.features.interactive = true`
 2. Worker starts task and listens for websocket connections to:
    `https://<hostname>:<port>/<secret>/shell`
 3. Worker publishes a redirect artifact called `private/docker/shell.html`
    which redirects to: 
    `https://tools.taskcluster.net/shell/?v=1&socketUrl=https://<hostname>:<port>/<secret>/shell`
 4. User clicks on the `private/docker/shell.html` artifact and is redirected to:
    `https://tools.taskcluster.net/shell/?v=1&socketUrl=https://<hostname>:<port>/<secret>/shell`
 5. Users browser reads `v` and `socketUrl=https://<hostname>:<port>/<secret>/shell`
    from querystring. If `v=1` the browser uses `docker-exec-websocket-server` to
    open a shell over a websocket. If `v=2` it uses `ws-shell` (to do the same thing).
 6. Users browser exposes the shell object to the user through a terminal emulator
    running in the browser.

How to make this work in taskcluster-cli:
 1. User creates a task with `task.payload.features.interactive = true`
 2. Worker starts task and listens for websocket connections to:
    `https://<hostname>:<port>/<secret>/shell`
 3. Worker publishes a redirect artifact called `private/docker/shell.html`
    which redirects to: 
    `https://tools.taskcluster.net/shell/?v=1&socketUrl=https://<hostname>:<port>/<secret>/shell`
 4. Users passes `taskId` to taskcluster-cli like: `taskcluster-cli shell <taskId>`
 5. Then `taskcluster-cli` does a GET request to `private/docker/shell.html`, but
    **without** following redirects. It then reads the `Location` header which is:
    `https://tools.taskcluster.net/shell/?v=1&socketUrl=https://<hostname>:<port>/<secret>/shell`.
    (In the future this step might be modified, perhaps we have a JSON artifact containing the `socketUrl`).
 6. If `v=2` in the querystring from the URL in the `Location` header then `taskcluster-cli` uses
    `shellclient.Dial(socketURL, "", true)` to obtain a shell object. See docs for the
    [shellclient package](https://godoc.org/github.com/taskcluster/taskcluster-worker/plugins/interactive/shellclient).
    If `v=1` then `taskcluster-cli` uses a library compatible with `docker-exec-websocket-server`,
    this needs to be written...
 7. The rest is setting the terminal to raw-mode, and piping data from the shell object
    to/from stdin/stdout/stderr. This is actually done already in:
    https://github.com/taskcluster/taskcluster-worker/blob/master/commands/shell/cmd.go

Steps 3-5 are subject to change. In `taskcluster-worker` for example the default
artifact name is `private/interactive/shell.html`. However, it also publishes a
JSON artifact call `private/interactive/sockets.json` that contains the `socketUrl` and `v` (version).
Obviously, it's key to separate these into different functions. One for finding the
version and socketUrl, a second function for obtaining a shell object, and a
third function for setting the terminal to raw-mode and piping the input/output.

Things to do:
 * Make a GET request without following redirects to `private/docker/shell.html`
   extract `v` and `socketUrl` variables from querystring from URL in `Location` header.
 * Implement a `docker-exec-websocket-server` compatible library in golang, just the client side.
 * Run terminal in raw mode...

It be worthwhile to start with `socketUrl` and `v` as arguments to the command.
And then later support detecting them from a `taskId`, by listing artifacts etc..
The hardest part here is wriging the `docker-exec-websocket-server` compatible client library.
To do this you probably need to run `docker-exec-websocket-server` locally, and develop against it.
The client library should probably expose an object compatible with the `engines.Shell` interface
from:
https://github.com/taskcluster/taskcluster-worker/blob/0a84683e2414646ec54beb141d7e55c868436b79/engines/sandbox.go#L6-L28

Just because, then it has the same methods as `shellclient` does... Making it easy to use either
one interchangably by defining an interface.

**Comment by jonasfj** [comment]
Oh, and the reason it's hard to implement a `docker-exec-websocket-server` client is that we actually do back-pressure.
Meaning if data arrives too fast we'll tell the other side to slow down. This is implemented with special messages... Checkout the code for `docker-exec-websocket-server`.

**Comment by t0xicCode** [comment]
Perfect! Thanks. I'll start planning my code.

**Comment by t0xicCode** [comment]
Closed by #54
